vault/identity_store_schema.go

```diff
@@ -0,0 +1,117 @@
+package vault
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+func identityStoreSchema() *memdb.DBSchema {
+	iStoreSchema := &memdb.DBSchema{
+		Tables: make(map[string]*memdb.TableSchema),
+	}
+
+	schemas := []func() *memdb.TableSchema{
+		identityTableSchema,
+		entityTableSchema,
+	}
+
+	for _, schemaFunc := range schemas {
+		schema := schemaFunc()
+		if _, ok := iStoreSchema.Tables[schema.Name]; ok {
+			panic(fmt.Sprintf("duplicate table name: %s", schema.Name))
+		}
+		iStoreSchema.Tables[schema.Name] = schema
+	}
+
+	return iStoreSchema
+}
+
+func identityTableSchema() *memdb.TableSchema {
+	return &memdb.TableSchema{
+		Name: "identities",
+		Indexes: map[string]*memdb.IndexSchema{
+			"id": &memdb.IndexSchema{
+				Name:   "id",
+				Unique: true,
+				Indexer: &memdb.StringFieldIndex{
+					Field: "ID",
+				},
+			},
+			"entity_id": &memdb.IndexSchema{
+				Name:   "entity_id",
+				Unique: false,
+				Indexer: &memdb.StringFieldIndex{
+					Field: "EntityID",
+				},
+			},
+			"mount_type": &memdb.IndexSchema{
```

Comment:
The items in the struct will always be at the top level. The question is on which items in the struct do we want to index the inserted object. Here we are indexing on mount_type because we want to be able to query all the identities of a specific type (per type and not per backend; example, all the github identities). Factors on the other hand, is used not for querying, but for ensuring the uniqueness of identities using a combination of mount_id and the identity_name. If we want to query identities of a specific mount, mount_id would be a separate index, at the top level.